src/main/java/com/lunark/lunark/properties/specification/PropertySpecification.java

```diff
@@ -0,0 +1,114 @@
+package com.lunark.lunark.properties.specification;
+
+import com.lunark.lunark.amenities.model.Amenity;
+import com.lunark.lunark.properties.dto.PropertySearchDto;
+import com.lunark.lunark.properties.model.Property;
+import com.lunark.lunark.properties.model.PropertyAvailabilityEntry;
+import jakarta.persistence.criteria.*;
+import org.springframework.data.jpa.domain.Specification;
+
+import java.time.LocalDate;
+
+public class PropertySpecification implements Specification<Property> {
+
+    private final PropertySearchDto filter;
+
+    public PropertySpecification(PropertySearchDto filter) {
+        super();
+        this.filter = filter;
+    }
+
+    @Override
+    public Predicate toPredicate(Root<Property> root, CriteriaQuery<?> query, CriteriaBuilder criteriaBuilder) {
```

Comment:
This method could have been broken down into several methods that generate Predicates which get connected in this method.